Fixes #27180 Provisioning host using vlan tagged network

vgunasek opened the pull request:
Using this modification we will be able to provision the host using VLAN tagged interface using both Host based ISO and Full Host Iso.

vgunasek committed:
Fixes #27180 Provisioning host using vlan tagged network

lzap commented on provisioning_templates/PXELinux/kickstart_default_pxelinux.erb:
```diff
@@ -32,12 +32,20 @@ oses:
     mask = subnet.mask
     gw = subnet.gateway
+    hostname = @host.name
+    interface = @host.provision_interface.identifier
```
Can you name this `identifier`?

lzap commented on provisioning_templates/PXELinux/kickstart_default_pxelinux.erb:
```diff
@@ -32,12 +32,20 @@ oses:
       # old Anacoda found in Fedora 16 or RHEL 6 and older
       dns_servers = subnet.dns_servers.join(',')
       options.push("ip=#{ip}", "netmask=#{mask}", "gateway=#{gw}", "dns=#{dns_servers}")
+      if @host.provision_interface.tag.present? &&  @host.provision_interface.attached_to.present?
```
Double space here.

lzap commented on provisioning_templates/PXELinux/kickstart_default_pxelinux.erb:
```diff
@@ -32,12 +32,20 @@ oses:
       dns_servers = subnet.dns_servers.join(',')
       options.push("ip=#{ip}", "netmask=#{mask}", "gateway=#{gw}", "dns=#{dns_servers}")
+      if @host.provision_interface.tag.present? &&  @host.provision_interface.attached_to.present?
+        options.push("vlanid=#{@host.provision_interface.tag} ksdevice=#{@host.provision_interface.identifier} rd.bootif=0")
```
Make use of the `identifier` variable here.

lzap commented on provisioning_templates/PXELinux/kickstart_default_pxelinux.erb:
```diff
@@ -32,12 +32,20 @@ oses:
     else
-      options.push("ip=#{ip}::#{gw}:#{mask}:::none")
+      options.push("ip=#{ip}::#{gw}:#{mask}:#{hostname}:#{interface}:none")
+      if @host.provision_interface.tag.present? &&  @host.provision_interface.attached_to.present?
```
Double space.

lzap commented on provisioning_templates/PXELinux/kickstart_default_pxelinux.erb:
```diff
@@ -32,12 +32,20 @@ oses:
-      options.push("ip=#{ip}::#{gw}:#{mask}:::none")
+      options.push("ip=#{ip}::#{gw}:#{mask}:#{hostname}:#{interface}:none")
+      if @host.provision_interface.tag.present? &&  @host.provision_interface.attached_to.present?
+        options.push("vlan=#{interface}:#{@host.provision_interface.attached_to} bootdev=#{interface} ifname=#{@host.provision_interface.attached_to}:#{mac} rd.bootif=0")
```
Use `identifier` instead of misleadingg `interface` here.

lzap commented on provisioning_templates/iPXE/kickstart_default_ipxe.erb:
```diff
@@ -16,13 +16,26 @@ oses:
+<%
+options = []
+hostname = @host.name
+interface = @host.provision_interface.identifier
```
Ditto. Or alternatively use `iface = @host.provision_interface` and then simply use `iface.identifier` everywhere?

lzap commented on provisioning_templates/iPXE/kickstart_default_ipxe.erb:
```diff
@@ -16,13 +16,26 @@ oses:
+      if @host.provision_interface.tag.present? &&  @host.provision_interface.attached_to.present?
+        options.push("vlanid=#{@host.provision_interface.tag} ksdevice=#{@host.provision_interface.identifier}")
+      end
+else
```
Correct indentation please.

lzap commented on provisioning_templates/provision/kickstart_default.erb:
```diff
@@ -80,7 +80,7 @@ skipx
 <% end -%>
 
-network --bootproto <%= dhcp ? 'dhcp' : "static --ip=#{@host.ip} --netmask=#{subnet.mask} --gateway=#{subnet.gateway} --nameserver=#{subnet.dns_servers.join(',')} --mtu=#{subnet.mtu.to_s}" %> --hostname <%= @host %><%= os_major >= 6 ? " --device=#{@host.mac}" : '' -%>
+network --bootproto <%= dhcp ? 'dhcp' : "static --ip=#{@host.ip} --netmask=#{subnet.mask} --gateway=#{subnet.gateway} --nameserver=#{subnet.dns_servers.join(',')} --mtu=#{subnet.mtu.to_s}" %> --hostname <%= @host %><%= os_major >= 6 ? " --device=#{@host.provision_interface.attached_to}" : '' -%> <%= (@host.provision_interface.tag.present? && @host.provision_interface.attached_to.present?) ? "--vl ...
```
I see a problem here. If attached_to is not set, previously we set this to "mac" but now it will be blank. You need to do if-statement too: `@host.provision_interface.tag.present? && @host.provision_interface.attached_to.present?`. However this line is getting out of control, I recommend to take similar approach as above - create an array, push into it and then join it with spaces. Take attention when editing, this must not break it's very important template :-)

lzap reviewed (changes requested):
Thanks a bunch. I wonder if it's better to define `iface = @host.provision_interface` and then simply use `iface.identifier` or `iface.xxxx` everywhere? The `interface` variable for identifier is a bit misleading, this is short enough.

Then there is one major concern around change on the kickstart "network" line. I suggest to carefully break it via array.

vgunasek committed:
Merge branch 'develop' of https://github.com/theforeman/community-templates into develop

vgunasek committed:
Have applied the changes as suggested:-

changes:-

interface --> iface

interface = @host.provision_interface.identifier
iface = @host.provision_interface

Provisioning template:-

<%- if iface.tag.present? &&  iface.attached_to.present?
  boot_options = "--device=#{iface.attached_to} --vlanid=#{iface.tag}"
else
  boot_options = os_major >= 6 ? " --device=#{@host.mac}" : ''
end -%>

boot_options..

lzap commented on provisioning_templates/PXELinux/kickstart_default_pxelinux.erb:
```diff
@@ -32,12 +32,20 @@ oses:
       # old Anacoda found in Fedora 16 or RHEL 6 and older
       dns_servers = subnet.dns_servers.join(',')
       options.push("ip=#{ip}", "netmask=#{mask}", "gateway=#{gw}", "dns=#{dns_servers}")
+      if iface.tag.present? &&  iface.attached_to.present?
```
```suggestion
      if iface.tag.present? && iface.attached_to.present?
```

lzap commented on provisioning_templates/PXELinux/kickstart_default_pxelinux.erb:
```diff
@@ -32,12 +32,20 @@ oses:
     else
-      options.push("ip=#{ip}::#{gw}:#{mask}:::none")
+      options.push("ip=#{ip}::#{gw}:#{mask}:#{hostname}:#{iface.identifier}:none")
+      if iface.tag.present? &&  iface.attached_to.present?
```
```suggestion
      if iface.tag.present? && iface.attached_to.present?
```

lzap commented on provisioning_templates/iPXE/kickstart_default_ipxe.erb:
```diff
@@ -16,21 +16,36 @@ oses:
 <%- end -%>
 
-kernel <%= "#{@host.url_for_boot(:kernel)}" %> initrd=initrd.img ks=<%= foreman_url('provision')%><%= static %> inst.stage2=<%= @host.operatingsystem.medium_uri(@host) %> <%= stage2 %> ksdevice=<%= @host.mac %> network kssendmac ks.sendmac inst.ks.sendmac <%= net_options %> <%= fips %>
+kernel <%= "#{@host.url_for_boot(:kernel)}" %> initrd=initrd.img ks=<%= foreman_url('provision')%><%= static %> inst.stage2=<%= @host.operatingsystem.medium_uri(@host) %> <%= stage2 %> ksdevice=<%= @host.mac %> network kssendmac ks.sendmac inst.ks.sendmac <%= ksoptions %> <%= fips %>
```
You can probably avoid introducing `ksoptions` variable and just do `options.join(' ')` on the `kernel ...` line directly.

vgunasek replied:
I have used ksoptions so that the same option will be used in other templates as well e.g kickstart_default_pxelinux.erb file. Can i proceed with your suggestion or can we keep it as it is?

lzap replied:
It's not big deal, up to you.

lzap commented on provisioning_templates/iPXE/kickstart_default_ipxe.erb:
```diff
@@ -16,21 +16,36 @@ oses:
+        end
+    end -%>
+
+<% ksoptions = options.join(' ')-%>
```
Missing space after `)`.

lzap commented on provisioning_templates/iPXE/kickstart_default_ipxe.erb:
```diff
@@ -16,21 +16,36 @@ oses:
-<%- else -%>
-<%- net_options = 'ip=${netX/ip}::${netX/gateway}:${netX/netmask}:::none nameserver=${dns}' -%>
-<%- end -%>
+<%- if (@host.operatingsystem.name.match(/Fedora/i) && @host.operatingsystem.major.to_i < 17) || @host.operatingsystem.major.to_i < 7
```
I suggest this style:

```erb
<%
  if (....)
  ...
-%>
```

lzap commented on provisioning_templates/provision/kickstart_default.erb:
```diff
@@ -43,7 +43,15 @@ This template accepts the following parameters:
+  iface = @host.provision_interface
 %>
+
+<%- if iface.tag.present? &&  iface.attached_to.present?
```
```suggestion
<%- if iface.tag.present? && iface.attached_to.present?
```

lzap commented on provisioning_templates/provision/kickstart_default.erb:
```diff
@@ -43,7 +43,15 @@ This template accepts the following parameters:
+  boot_options = "--device=#{iface.attached_to} --vlanid=#{iface.tag}"
+else
+  boot_options = os_major >= 6 ? " --device=#{@host.mac}" : ''
+end -%>
```
I suggest to do similar approach as above - make `boot_options` an array and then do `boot_options.join(' ')` on the `network` line. Probably the variable should be named something like `network_options` which fits better.

lzap reviewed (changes requested):
Couple of nitpicks but overall it looks great.

vgunasek committed:
Merge branch 'develop' of https://github.com/theforeman/community-templates into develop

vgunasek committed:
Update the template as per lzap suggestions.

lzap commented on provisioning_templates/provision/kickstart_default.erb:
```diff
@@ -43,7 +43,16 @@ This template accepts the following parameters:
+<%- if iface.tag.present? && iface.attached_to.present?
+  network_options.push = "--device=#{iface.attached_to} --vlanid=#{iface.tag}"
+else
+  network_options.push = os_major >= 6 ? " --device=#{@host.mac}" : ''
```
This and line 51 are syntax errors, `push` is a method which you call. Remove the equal sign.

```suggestion
  network_options.push(os_major >= 6 ? " --device=#{@host.mac}" : '')
```

lzap commented on provisioning_templates/provision/kickstart_default.erb:
```diff
@@ -43,7 +43,16 @@ This template accepts the following parameters:
 %>
+
+<%- if iface.tag.present? && iface.attached_to.present?
+  network_options.push = "--device=#{iface.attached_to} --vlanid=#{iface.tag}"
```
Also here, syntax error.

```suggestion
  network_options.push("--device=#{iface.attached_to} --vlanid=#{iface.tag}")
```

lzap reviewed (changes requested):
There are two syntax errors there. To test this you can run template render test from a foreman foldere:

```
bundle exec rake templates:render DIRECTORY=../community-templates/provisioning_templates/

 ==== Kickstart default ====
(../community-templates/provisioning_templates/provision/kickstart_default.erb)
 -> Safemode doesn't allow to access 'attrasgn' on network_options.push = "--device=#{iface.to_jail.attached_to} --vlanid=#{iface.to_jail.tag}"

 ==== Kickstart default iPXE ====
(../community-templates/provisioning_templates/iPXE/kickstart_default_ipxe.erb)
 -> Safemode doesn't allow to access 'interface' on #<Safemode::ScopeObject>
```

Errors are a bit weird to understand (you need to know a lot about safemode), but it must pass without any of them.

lzap commented on provisioning_templates/iPXE/kickstart_default_ipxe.erb:
```diff
@@ -16,21 +16,38 @@ oses:
+      options.push("vlanid=#{iface.tag} ksdevice=#{iface.identifier}")
+    end
+  else
+    options.push("ip=${netX/ip}::${netX/gateway}:${netX/netmask}:#{hostname}:#{interface}:none nameserver=${dns}")
```
Here is the second error: `interface` is not known variable. You mean `iface` instead.

vgunasek committed:
Merge branch 'develop' of https://github.com/theforeman/community-templates into develop

mmoll commented:
what's the status here?

lzap commented:
I finally have time to test this. It does not work, there are some issues and also tests are failing. Let me close this one and then open a new PR where I put my changes on top of your changes:

https://github.com/theforeman/community-templates/pull/641

Can you take a look and test those templates, specifically iPXE since you already have the setup? I tested grub2 already.

vgunasek commented:
Okay sure  buddy i am deploying my foreman server will test and share the test result.
